mh_sha1/mh_sha1_block_avx512.asm

```diff
@@ -293,7 +280,7 @@ func(mh_sha1_block_avx512)
 	vmovdqa64  D, HH3
 	vmovdqa64  E, HH4
 
-	vmovdqa32	KT, [K00_19]
+	vpbroadcastd	KT, [rax+K00_19-PSHUFFLE_BYTE_FLIP_MASK]
```

Comment:
It's true that we could load alla the data using RIP relative addressing but it makes the (already 6-bytes long) instructions even longer thanks to the 32bit immediate. Given that all the constants are not far apart using rax as a base register into the .data section helps reducing the code size. Looks more complex but if you look at the code in a disassembler it's actually a simple register+offset addressing.

Reply:
I do believe opcodes are smaller but does it give any other benefit than this? The code is more difficult to read this way.

Reply:
I can add some comments and label at the same address in the .data section with something like "base" as a name so it's clearer that it's just a way to make the assembler calculate the offset from the base address in the .data section?

Reply:
Thanks for exploring the alternatives. The point is that if such change is not bringing any benefits, i.e. performance improvement, and it makes the code more difficult to maintain then it is hard to justify making the change.

With replacement of moves with broadcasts, I can see benefit of reduced memory footprint. It may help in larger applications and it shouldn't affect algorithm performance.

Can benefit of having smaller opcodes be quantified in some way?

Reply:
I agree that it's not easy to quantify the difference in performance between RIP addressing vs register+offset (beside the shorter instructions). It's probably micro-architecture dependent. Some micro-archs have higher latency or less throughtput for RIP addressing mode, same goes for the amount of slots in the mu-ops cache that can store a wide immediate.

Give all of this i went for a simple base+offset addressing mode, because IMO we can be reasonably sure that it will always be implemented as a "fast" addressing mode in any reasonable x86 implementation. The shorter encoding is also a bonus of course. The code is a bit more complex to read, but there're much more opaque macros in the repository IMO but seems like a good tradeoff to achive an optimal (size and speed) implementation. All of this IMHO of course, but the beauty of assembly is that we can do this kind of tricks to really polish the code :)

Reply:
OK thanks for taking time to reply. I am thinking to limit this pull request to use of broadcasts and reducing memory footprint (clear improvement). We can get back to other ideas in the future (i.e. optimizing front end operations).
Could you re-work this PR and keep only the broadcast changes?

Thanks again for submitting the pull request and contributing to isa-l_crypto 👍

Reply:
Sure, so only the broadcast or i can leave minor thing like keeping the shuffle mask in registers instead of reloading it every time (in the sha256 version)?. I'll try to rework this PR with only that stuff, i'm still learning all the tricks of git.
Thanks again for the review and taking time to explain your rationale 😄